dice_ml/explainer_interfaces/dice_KD.py

```diff
@@ -0,0 +1,236 @@
+"""
+Module to generate diverse counterfactual explanations based on tensorflow 2.x
+"""
+from dice_ml.explainer_interfaces.explainer_base import ExplainerBase
+import tensorflow as tf
+
+import numpy as np
+import random
+import timeit
+from sklearn.neighbors import KDTree
+import pandas as pd
+
+from dice_ml import diverse_counterfactuals as exp
+
+class DiceKD(ExplainerBase):
+
+    def __init__(self, data_interface, model_interface):
+        """Init method
+
+        :param data_interface: an interface class to access data related params.
+        :param model_interface: an interface class to access trained ML model.
+
+        """
+
+        super().__init__(data_interface) # initiating data related parameters
+
+        # initializing model variables
+        self.model = model_interface
+
+        # loading trained model
+        self.model.load_model()
+
+        # variables required to generate CFs - see generate_counterfactuals() for more info
+        self.cfs = []
+        self.features_to_vary = []
+        self.cf_init_weights = []  # total_CFs, algorithm, features_to_vary
+        self.loss_weights = []  # yloss_type, diversity_loss_type, feature_weights
+        self.feature_weights_input = ''
+        self.hyperparameters = [1, 1, 1]  # proximity_weight, diversity_weight, categorical_penalty
+        self.optimizer_weights = []  # optimizer, learning_rate
+
+        # number of output nodes of ML model
+        temp_input = tf.convert_to_tensor([tf.random.uniform([len(self.data_interface.encoded_feature_names)])], dtype=tf.float32)
+        self.num_ouput_nodes = self.model.get_output(temp_input).shape[1]
+
+        # Stores the predictions on the training data
+        #TODO: Implement this for private dataset instance too
+        dataset_dict = self.data_interface.prepare_dataset_instance(encode=True)
+        dataset_dict_output = np.array([dataset_dict.values], dtype=np.float32)
+
+        self.predictions = self.predict_fn(dataset_dict_output)
+        self.predictions_vals = np.reshape(self.predictions[0], (self.predictions.shape[1],))
+
+        #segmenting the dataset according to outcome
```

Comment:
init is usually only for initialization. the code to build kd trees (after line 54) should ideally come under a separate function, `build_kd_trees` that you simply call from __init__.